design/user-event-spi.md

```diff
@@ -0,0 +1,149 @@
+# User Event SPI
+
+* **Status**: Draft v0.1
+* **Author**: [vmuzikar](https://github.com/vmuzikar)
+
+## Motivation
+
+Current Event SPI design has several limitations. E.g. it's not defined when an event is fired exactly, there are no pre-
+and post-events. There's also no way of pre-events intercepting for approve/reject functionality.
+
+## Use cases
+
+* A user is removed from Keycloak. All related records in an external system are removed before the user is permanently
+  deleted (e.g. eshop orders).
+* A user self-registers in Keycloak. Their registration must be approved by an external system (either human or automated)
+  before the new account is actually fully created.
+* A user self-edits their email address. The change must be approved first by an external system before it's actually
+  applied.
+
+## Design proposal
+
+To overcome the current Event SPI limitations, a new "domain-specific SPI" will be implemented – User Event SPI.
+
+Key changes over current generic Event SPI:
+* Each event is fired twice – as a pre- and post-event.
+* Pre-events are "interactive" – listener can approve, reject or delegate (for later decision by an external system)
+  each pre-event. 
+* The same User Event SPI will be used everywhere – admin console and REST endpoints, user registration form,
+  account console and REST endpoints, etc.
+* The event types will be much more detailed, not just usual CRUD. There'll be separated events for user self-registration,
+  creation through Admin REST API, Account Console, etc.
+
+### Terminology
+
+User Action (in this context) is a process which fires the User Event. This can be e.g. a user creation through Admin REST
+API, user self-registration, user field modification, assigning a user role, etc.
+
+Pre-event is an event that is fired after all other verifications passed (authorization, request validation, etc.), just
+before persisting any changes.
+
+Post-event is an event that is fired after a User Action has finished completely and all changes are persisted.
+
+### Event types
+
+* User is created manually through admin endpoints
+* Basic user info (name, email, etc.) is modified through admin endpoints
+* Basic user info (name, email, etc.) is modified through account endpoints
+* User self-registers through login page
+* User self-registers through an IdP, i.e. logs in for the first time
+* Admin adds a federation link
+* Admin removes a federation link
+* User logs in using federation for the first time
```

Comment:
How about when the user is created via federation? E.g. when a user is added to the company LDAP, it is synchronised to the relevant Keycloak Realms. From there, the Event SPI could be used to synchronise to another downstream system.
This would be very useful for syncing users to other systems, e.g. a HR system

Reply:
> From there, the Event SPI could be used to synchronise to another downstream system.
This would be very useful for syncing users to other systems, e.g. a HR system

Isn't that something that belongs in an IDM system (Midpoint IDM, Wren:IDM, etc.) instead of an AM system like Keycloak?

Reply:
@jonm-bb Technically speaking it'd be possible. The use case you're describing is exactly what I had in mind by "User logs in using federation for the first time". In other words, let's say a user exists in some LDAP; then the user logs in for the first time to Keycloak (i.e. user is created in Keycloak via federation); then an event is fired.

Reply:
@siepkes - sure, you could involve more products and systems, but I think this kind of SPI makes integrating with such systems easier, and could even avoid then need for them for simple deployments. If it doesn't add significantly to the complexity, then it would add quite a lot of value to the product.
@vmuzikar - ah cool, makes sense. Could we clarify that the event is fired when the user is persisted as well as when they log in?

Reply:
@jonm-bb Current design includes only events for when the user is persisted.

Reply:
> Current design includes only events for when the user is persisted.
Okay, then so this line:
> User logs in using federation for the first time
I think we could change this to "User record is created via federation"
Does that make sense?